<a href="https://colab.research.google.com/github/olgaoblomova/Python_repository/blob/main/Final_work_Metrics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('Online Retail_.csv', dtype = {'CustomerID':str})

In [ ]:
df

InvoiceNo StockCode  ... CustomerID         Country
0        536365    84029E  ...      17850  United Kingdom
1        536367     84879  ...      13047  United Kingdom
2        536367     21755  ...      13047  United Kingdom
3        536368     22914  ...      13047  United Kingdom
4        536373    84406B  ...      17850  United Kingdom
...         ...       ...  ...        ...             ...
54209    581585     84832  ...      15804  United Kingdom
54210    581585     23084  ...      15804  United Kingdom
54211    581585     84692  ...      15804  United Kingdom
54212    581587     22631  ...      12680          France
54213    581587     23255  ...      12680          France

[54214 rows x 8 columns]

In [ ]:
df.value_counts('Country').count()

37

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54214 entries, 0 to 54213
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   InvoiceNo    54214 non-null  object 
 1   StockCode    54214 non-null  object 
 2   Description  54080 non-null  object 
 3   Quantity     54214 non-null  int64  
 4   InvoiceDate  54214 non-null  object 
 5   UnitPrice    54214 non-null  float64
 6   CustomerID   40643 non-null  object 
 7   Country      54214 non-null  object 
dtypes: float64(1), int64(1), object(6)
memory usage: 3.3+ MB


Имеем незаполненные данные по столбцам Description и CustomerID.  Необходимо преобразовать столбец с датой в формат даты

In [ ]:
from datetime import datetime

In [ ]:
df['InvoiceDate'] = df['InvoiceDate'].apply(lambda d: datetime.strptime(d, '%m/%d/%y %H:%M'))

In [ ]:
df

InvoiceNo StockCode  ... CustomerID         Country
0        536365    84029E  ...      17850  United Kingdom
1        536367     84879  ...      13047  United Kingdom
2        536367     21755  ...      13047  United Kingdom
3        536368     22914  ...      13047  United Kingdom
4        536373    84406B  ...      17850  United Kingdom
...         ...       ...  ...        ...             ...
54209    581585     84832  ...      15804  United Kingdom
54210    581585     23084  ...      15804  United Kingdom
54211    581585     84692  ...      15804  United Kingdom
54212    581587     22631  ...      12680          France
54213    581587     23255  ...      12680          France

[54214 rows x 8 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54214 entries, 0 to 54213
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   InvoiceNo    54214 non-null  object        
 1   StockCode    54214 non-null  object        
 2   Description  54080 non-null  object        
 3   Quantity     54214 non-null  int64         
 4   InvoiceDate  54214 non-null  datetime64[ns]
 5   UnitPrice    54214 non-null  float64       
 6   CustomerID   40643 non-null  object        
 7   Country      54214 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 3.3+ MB


In [ ]:
df.duplicated('InvoiceNo')

0        False
1        False
2         True
3        False
4        False
         ...  
54209     True
54210     True
54211     True
54212    False
54213     True
Length: 54214, dtype: bool

In [ ]:
df_null_Descr = df[df['Description'].isnull()]
df_null_Descr

InvoiceNo StockCode Description  ...  UnitPrice CustomerID         Country
204      536553     37461         NaN  ...        0.0        NaN  United Kingdom
730      536995     35951         NaN  ...        0.0        NaN  United Kingdom
731      536998     85067         NaN  ...        0.0        NaN  United Kingdom
732      537005     21692         NaN  ...        0.0        NaN  United Kingdom
733      537009    84534B         NaN  ...        0.0        NaN  United Kingdom
...         ...       ...         ...  ...        ...        ...             ...
48602    577608     35400         NaN  ...        0.0        NaN  United Kingdom
49770    578360    84971L         NaN  ...        0.0        NaN  United Kingdom
50021    578693     21811         NaN  ...        0.0        NaN  United Kingdom
52171    580359     20775         NaN  ...        0.0        NaN  United Kingdom
53537    581209     21620         NaN  ...        0.0        NaN  United Kingdom

[134 rows x 8 columns]

In [ ]:
df_null_Descr[df_null_Descr['UnitPrice']!=0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [ ]:
df[df['Description'].isnull()].count()

InvoiceNo      134
StockCode      134
Description      0
Quantity       134
InvoiceDate    134
UnitPrice      134
CustomerID       0
Country        134
dtype: int64

134 записи в столбце Description не содержат данных

Строки с пустыми значениями по столбцу Description имеют также пустые значения в поле цена UnitPrice. Принимаю решение удалить их (0,25 % от объема данных)

In [ ]:
df.dropna(subset = ['Description'], inplace=True)

In [ ]:
df

InvoiceNo StockCode  ... CustomerID         Country
0        536365    84029E  ...      17850  United Kingdom
1        536367     84879  ...      13047  United Kingdom
2        536367     21755  ...      13047  United Kingdom
3        536368     22914  ...      13047  United Kingdom
4        536373    84406B  ...      17850  United Kingdom
...         ...       ...  ...        ...             ...
54209    581585     84832  ...      15804  United Kingdom
54210    581585     23084  ...      15804  United Kingdom
54211    581585     84692  ...      15804  United Kingdom
54212    581587     22631  ...      12680          France
54213    581587     23255  ...      12680          France

[54080 rows x 8 columns]

In [ ]:
df[df['Description'].isnull()]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [ ]:
df[df['CustomerID'].isnull()]

InvoiceNo StockCode  ... CustomerID         Country
158      536544     21809  ...        NaN  United Kingdom
159      536544     21870  ...        NaN  United Kingdom
160      536544     21879  ...        NaN  United Kingdom
161      536544     21928  ...        NaN  United Kingdom
162      536544     22069  ...        NaN  United Kingdom
...         ...       ...  ...        ...             ...
54161    581498     48138  ...        NaN  United Kingdom
54162    581498     82482  ...        NaN  United Kingdom
54163    581498    84032A  ...        NaN  United Kingdom
54164    581498    84558a  ...        NaN  United Kingdom
54165    581498     85150  ...        NaN  United Kingdom

[13437 rows x 8 columns]

In [ ]:
df[df['CustomerID'].isnull()].count()

InvoiceNo      13437
StockCode      13437
Description    13437
Quantity       13437
InvoiceDate    13437
UnitPrice      13437
CustomerID         0
Country        13437
dtype: int64

25% от первоначальных данных по столбцу CustomerID не заполнено, предполагаю, что покупки совершены незарегистрированными пользователями, заполню пропуски значением unregistered



In [ ]:
df['CustomerID'].fillna('unregistered', inplace=True)

In [ ]:
df

InvoiceNo StockCode  ... CustomerID         Country
0        536365    84029E  ...      17850  United Kingdom
1        536367     84879  ...      13047  United Kingdom
2        536367     21755  ...      13047  United Kingdom
3        536368     22914  ...      13047  United Kingdom
4        536373    84406B  ...      17850  United Kingdom
...         ...       ...  ...        ...             ...
54209    581585     84832  ...      15804  United Kingdom
54210    581585     23084  ...      15804  United Kingdom
54211    581585     84692  ...      15804  United Kingdom
54212    581587     22631  ...      12680          France
54213    581587     23255  ...      12680          France

[54080 rows x 8 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 54080 entries, 0 to 54213
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   InvoiceNo    54080 non-null  object        
 1   StockCode    54080 non-null  object        
 2   Description  54080 non-null  object        
 3   Quantity     54080 non-null  int64         
 4   InvoiceDate  54080 non-null  datetime64[ns]
 5   UnitPrice    54080 non-null  float64       
 6   CustomerID   54080 non-null  object        
 7   Country      54080 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 3.7+ MB


In [ ]:
df[df['UnitPrice']<0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [ ]:
df[df['UnitPrice']==0]

InvoiceNo StockCode  ...    CustomerID         Country
1424     537534     85064  ...  unregistered  United Kingdom
1425     537534     22681  ...  unregistered  United Kingdom
1426     537534     22520  ...  unregistered  United Kingdom
1427     537534     22371  ...  unregistered  United Kingdom
1428     537534     21912  ...  unregistered  United Kingdom
...         ...       ...  ...           ...             ...
51407    579563    85135C  ...  unregistered  United Kingdom
51498    579732     35648  ...  unregistered  United Kingdom
51508    579743     21810  ...  unregistered  United Kingdom
53536    581207     21688  ...  unregistered  United Kingdom
53702    581226     23090  ...  unregistered  United Kingdom

[91 rows x 8 columns]

Как видим 91 строка с нулевыми ценами товаров еще осталась, кроме того в столбце Description ест строки где вместо описания товара посторонние записи и минусовые цены. Удаляю.

In [ ]:
df.drop(df[df['UnitPrice']==0].index, axis = 0, inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 53989 entries, 0 to 54213
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   InvoiceNo    53989 non-null  object        
 1   StockCode    53989 non-null  object        
 2   Description  53989 non-null  object        
 3   Quantity     53989 non-null  int64         
 4   InvoiceDate  53989 non-null  datetime64[ns]
 5   UnitPrice    53989 non-null  float64       
 6   CustomerID   53989 non-null  object        
 7   Country      53989 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 3.7+ MB


In [ ]:
df[df['UnitPrice']==0].count()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [ ]:
df[df['Quantity']<0]

InvoiceNo StockCode  ... CustomerID         Country
19      C536383    35004C  ...      15311  United Kingdom
30      C536391     21983  ...      17548  United Kingdom
31      C536391     21980  ...      17548  United Kingdom
202     C536548     20957  ...      12472         Germany
375     C536642     21463  ...      14390  United Kingdom
...         ...       ...  ...        ...             ...
53704   C581229    79191C  ...      12558             USA
53705   C581229     21507  ...      12558             USA
53765   C581322     22666  ...      18223  United Kingdom
53770   C581330     22959  ...      15877  United Kingdom
53998   C581465     22171  ...      15755  United Kingdom

[960 rows x 8 columns]

In [ ]:
df[df['Quantity']<0].count()

InvoiceNo      960
StockCode      960
Description    960
Quantity       960
InvoiceDate    960
UnitPrice      960
CustomerID     960
Country        960
dtype: int64

960 строки имеют отрицательное значение в столбце Quantity

Вероятно данные по количеству некорректно записывались.Возможно так записывались возвраты, но проанализировав несколько строк по ID покупателя, не нашла изначально проданных этому покупателю этих наименований товара. Принимаю решение поменять знак на положительный в этих строках.

In [ ]:
def change_sign(d):
  if d > 0:
    return d
  else:
    return d*(-1)

In [ ]:
df['Quantity'] = df['Quantity'].apply(change_sign)

In [ ]:
df

InvoiceNo StockCode  ... CustomerID         Country
0        536365    84029E  ...      17850  United Kingdom
1        536367     84879  ...      13047  United Kingdom
2        536367     21755  ...      13047  United Kingdom
3        536368     22914  ...      13047  United Kingdom
4        536373    84406B  ...      17850  United Kingdom
...         ...       ...  ...        ...             ...
54209    581585     84832  ...      15804  United Kingdom
54210    581585     23084  ...      15804  United Kingdom
54211    581585     84692  ...      15804  United Kingdom
54212    581587     22631  ...      12680          France
54213    581587     23255  ...      12680          France

[53989 rows x 8 columns]

In [ ]:
df[df['Quantity']<0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [ ]:
df.describe()

Quantity     UnitPrice
count  53989.000000  53989.000000
mean      10.489692      4.325292
std       38.294949     47.933260
min        1.000000      0.001000
25%        1.000000      1.250000
50%        3.000000      2.100000
75%       10.000000      4.130000
max     2400.000000   6930.000000

In [ ]:
df

InvoiceNo StockCode  ... CustomerID         Country
0        536365    84029E  ...      17850  United Kingdom
1        536367     84879  ...      13047  United Kingdom
2        536367     21755  ...      13047  United Kingdom
3        536368     22914  ...      13047  United Kingdom
4        536373    84406B  ...      17850  United Kingdom
...         ...       ...  ...        ...             ...
54209    581585     84832  ...      15804  United Kingdom
54210    581585     23084  ...      15804  United Kingdom
54211    581585     84692  ...      15804  United Kingdom
54212    581587     22631  ...      12680          France
54213    581587     23255  ...      12680          France

[53989 rows x 8 columns]

Очищенный файл выгружаю в csv-формат для построения дашборда""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

In [ ]:
df.to_csv("output.csv", index = False, line_terminator ='\n',)

Проверю корректность выгруженных данных

In [ ]:
dff = pd.read_csv('output.csv')

In [ ]:
dff

InvoiceNo StockCode  ... CustomerID         Country
0        536365    84029E  ...      17850  United Kingdom
1        536367     84879  ...      13047  United Kingdom
2        536367     21755  ...      13047  United Kingdom
3        536368     22914  ...      13047  United Kingdom
4        536373    84406B  ...      17850  United Kingdom
...         ...       ...  ...        ...             ...
53984    581585     84832  ...      15804  United Kingdom
53985    581585     23084  ...      15804  United Kingdom
53986    581585     84692  ...      15804  United Kingdom
53987    581587     22631  ...      12680          France
53988    581587     23255  ...      12680          France

[53989 rows x 8 columns]

In [ ]:
dff.describe()

Quantity     UnitPrice
count  53989.000000  53989.000000
mean      10.489692      4.325292
std       38.294949     47.933260
min        1.000000      0.001000
25%        1.000000      1.250000
50%        3.000000      2.100000
75%       10.000000      4.130000
max     2400.000000   6930.000000

Для предварительной оценки показателей добавлю столбец с итоговой суммой

In [ ]:
df['total_price']=df.apply(lambda d: d['UnitPrice']*d['Quantity'], axis=1)

In [ ]:
df['total_price'].sum()

1112240.091

In [ ]:
df.groupby('Country').sum('total_price').sort_values('total_price',  ascending=False)

Quantity   UnitPrice  total_price
Country                                                
United Kingdom          469211  208411.791   939628.821
EIRE                     14765    5489.260    30574.080
Netherlands              20943     824.160    26710.580
Germany                  12443    4384.710    24893.120
France                   11742    2929.530    20149.970
Australia                10045     334.340    17240.220
Switzerland               3543     607.900     6222.500
Sweden                    4680     211.880     6124.520
Spain                     2938     912.480     5986.860
Portugal                  1832    1821.550     4557.790
Belgium                   2406     764.920     4101.260
Norway                    1798    1096.410     3859.900
Hong Kong                  374    2675.050     3462.770
Channel Islands           1567     431.670     2877.950
Japan                     1594      72.090     2326.780
Denmark                    742     113.990     2058.840
Finland                   1017     299.850     2040.600
Italy                      766     333.550     1512.320
Cyprus                     852     503.620     1378.640
Singapore                  374     297.230     1045.680
Austria                    449     161.840     1036.270
Poland                     469     165.090      918.250
Israel                     331     193.750      531.720
Unspecified                292     126.910      520.220
Greece                     158      49.460      490.820
USA                        282      61.350      479.620
Malta                       46      76.300      293.700
Lithuania                  104      19.300      291.600
Iceland                    161      45.730      271.790
Canada                     226      18.770      214.140
United Arab Emirates        38      18.030       96.420
Czech Republic              36       4.900       82.200
European Community          33      26.500       80.900
Brazil                      28      18.450       79.800
RSA                         29       7.920       42.140
Lebanon                      8       4.950       39.600
Bahrain                      6       2.950       17.700

In [ ]:
df_invoice = df.groupby(['InvoiceNo']).sum('total_price').reset_index()

In [ ]:
df_invoice

InvoiceNo  Quantity  UnitPrice  total_price
0        536365         6       3.39        20.34
1        536367        35       7.64        71.93
2        536368         3       4.95        14.85
3        536373         8       2.75        22.00
4        536375         6       2.55        15.30
...         ...       ...        ...          ...
14675   C581228        24       6.75       162.00
14676   C581229        36       1.27        25.44
14677   C581322         1       2.95         2.95
14678   C581330        25       0.42        10.50
14679   C581465         1       8.50         8.50

[14680 rows x 4 columns]

In [ ]:
df_invoice = df_invoice[['InvoiceNo','total_price']]

Объем продаж за все время

In [ ]:
df_invoice['total_price'].sum()

1112240.091

Количество чеков за все время

In [ ]:
df_invoice['total_price'].count()

14680

Средний чек

In [ ]:
df_invoice['total_price'].mean()

75.76567377384171

In [ ]:
df_invoice.to_csv("output_invoice.csv", index = False, line_terminator ='\n',)

In [ ]:
df.groupby(['CustomerID']).sum()

Quantity  UnitPrice  total_price
CustomerID                                    
12347              161      45.73       271.79
12348              144       1.80       163.20
12349               83      17.50       106.84
12350               36       4.15        49.80
12352               26     110.00       184.40
...                ...        ...          ...
18281               10       1.65        16.50
18282                9       2.55        22.95
18283              148     116.46       215.30
18287              124      11.93       158.88
unregistered     43671   92295.86    187861.16

[3820 rows x 3 columns]

In [ ]:
df_quanitity_position = df[df['CustomerID']!='unregistered'].groupby(['CustomerID','InvoiceDate']).count()
df_quanitity_position

InvoiceNo  StockCode  ...  Country  total_price
CustomerID InvoiceDate                                ...                      
12347      2010-12-07 14:57:00          4          4  ...        4            4
           2011-01-26 14:30:00          2          2  ...        2            2
           2011-04-07 10:43:00          3          3  ...        3            3
           2011-10-31 12:25:00          4          4  ...        4            4
           2011-12-07 15:52:00          2          2  ...        2            2
...                                   ...        ...  ...      ...          ...
18283      2011-11-23 13:27:00         12         12  ...       12           12
           2011-11-30 12:59:00          4          4  ...        4            4
           2011-12-06 12:02:00          2          2  ...        2            2
18287      2011-05-22 10:39:00          4          4  ...        4            4
           2011-10-12 10:23:00          2          2  ...        2            2

[13660 rows x 7 columns]

Среднее количество позиций в чеке среди зарегистрированных пользователей

In [ ]:
df_quanitity_position['InvoiceNo'].mean()

2.9750366032210835